From bf2b2c1fc19545085295a60958bf66edc2a90ee7 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:17:37 -0400
Subject: [net] netfilter: nf_tables: reset rule number counter after jump and goto

Message-id: <20141021161737.22888.18170.stgit@localhost.localdomain>
Patchwork-id: 97941
O-Subject: [RHEL7 net PATCH 043/124] netfilter: nf_tables: reset rule number counter after jump and goto
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: d088be8042841f024156ee68fecfef7503d660cb

 commit d088be8042841f024156ee68fecfef7503d660cb
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Sat May 10 13:39:21 2014 +0200

    netfilter: nf_tables: reset rule number counter after jump and goto

    Otherwise we start incrementing the rule number counter from the
    previous chain iteration.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_core.c b/net/netfilter/nf_tables_core.c
index f4340dc..8b02202 100644
--- a/net/netfilter/nf_tables_core.c
+++ b/net/netfilter/nf_tables_core.c
@@ -123,7 +123,7 @@ nft_do_chain(struct nft_pktinfo *pkt, const struct nf_hook_ops *ops)
  struct nft_data data[NFT_REG_MAX + 1];
  unsigned int stackptr = 0;
  struct nft_jumpstack jumpstack[NFT_JUMP_STACK_SIZE];
- int rulenum = 0;
+ int rulenum;
  /*
   * Cache cursor to avoid problems in case that the cursor is updated
   * while traversing the ruleset.
@@ -131,6 +131,7 @@ nft_do_chain(struct nft_pktinfo *pkt, const struct nf_hook_ops *ops)
  unsigned int gencursor = ACCESS_ONCE(chain->net->nft.gencursor);
 
 do_chain:
+ rulenum = 0;
  rule = list_entry(&chain->rules, struct nft_rule, list);
 next_rule:
  data[NFT_REG_VERDICT].verdict = NFT_CONTINUE;
-- 
1.7.1